### PR TITLE
Fix resolving of root / URIs

### PR DESCRIPTION
**Pull request by armanbilge**
First two commits are just some tidying. The fix is in https://github.com/http4s/http4s/commit/2096b29f1f9d54a932ba161db1288a8dd106edd9.

### CODE REVIEW

**Comment by armanbilge** [comment]
Btw, this case isn't included in the examples in the spec. Any insight?

https://www.rfc-editor.org/rfc/rfc3986#section-5.4.1

**Comment by kubukoz** [comment]
now _that_ is a scary looking diff. Compared to what it really is 😬

**Review by rossabaker — APPROVED** [review]
No insight on the missing case in the RFC, but it's intuitive and matches `java.net.URI` and Python's `urllib.parse.urljoin`.  The former implements an old RFC and the latter is another language, but at least it's not counterevidence.

File: core/shared/src/main/scala/org/http4s/Uri.scala [hunk 1]
```diff
@@ -929,8 +929,10 @@ object Uri extends UriPlatform {
     val target = (base, reference) match {
       case (_, Uri(Some(_), _, _, _, _)) => reference
       case (Uri(s, _, _, _, _), Uri(_, a @ Some(_), p, q, f)) => Uri(s, a, p, q, f)
-      case (Uri(s, a, p, q, _), Uri(_, _, pa, Query.empty, f)) if pa.isEmpty => Uri(s, a, p, q, f)
-      case (Uri(s, a, p, _, _), Uri(_, _, pa, q, f)) if pa.isEmpty => Uri(s, a, p, q, f)
+      case (Uri(s, a, p, q, _), Uri(_, _, pa, Query.empty, f)) if pa.isEmpty && !pa.absolute =>
```
Comment (danicheg) [inline core/shared/src/main/scala/org/http4s/Uri.scala:932]: _a skippable minor grumpy nitpick._: does `pa.isEmpty && !pa.absolute` merely correspond to `pa == Path.empty`?

**Review by danicheg — APPROVED** [review]

File: core/shared/src/main/scala/org/http4s/Uri.scala (hunk 1 above)
Comment (armanbilge) [reply core/shared/src/main/scala/org/http4s/Uri.scala:932]: Hmm, that's a good question. It also opens the question that should `pa.isEmpty` be equivalent to `pa == Path.empty`? Is `/` empty or not? 😵

Comment (danicheg) [reply core/shared/src/main/scala/org/http4s/Uri.scala:932]: It's interesting, maybe, it's kinda side effect that `Path.empty` equals `/`.

**Pull request merged**